api/v1alpha1/commons.go

```diff
@@ -175,6 +175,68 @@ type VaultSecretReference struct {
+	// password: Specifies the password to use when connecting with the username. This value will not be returned by Vault when performing a read upon the configuration. This is typically used in the connection_url field via the templating directive "{{"password"}}".
+	// If username is provided as spec.username, it takes precedence over the username retrieved from the referenced secret
+	// +kubebuilder:validation:Optional
+	Secret *corev1.LocalObjectReference `json:"secret,omitempty"`
+
+	// RandomSecret retrieves the credentials from the Vault secret corresponding to this RandomSecret. This will map the "username" and "password" keys of the secret to the username and password of this config. All other keys will be ignored. If the RandomSecret is refreshed the operator retrieves the new secret from Vault and updates this configuration. Only one of RootCredentialsFromVaultSecret or RootCredentialsFromSecret or RootCredentialsFromRandomSecret can be specified.
+	// When using randomSecret a username must be specified in the spec.username
+	// password: Specifies the password to use when connecting with the username. This value will not be returned by Vault when performing a read upon the configuration. This is typically used in the connection_url field via the templating directive "{{"password"}}"".
+	// +kubebuilder:validation:Optional
+	RandomSecret *corev1.LocalObjectReference `json:"randomSecret,omitempty"`
+
+	// PasswordKey key to be used when retrieving the password, required with VaultSecrets and Kubernetes secrets, ignored with RandomSecret
+	// +kubebuilder:validation:Optional
+	PasswordKey string `json:"passwordKey,omitempty"`
+
+	// UsernameKey key to be used when retrieving the username, optional with VaultSecrets and Kubernetes secrets, ignored with RandomSecret
+	// +kubebuilder:validation:Optional
+	UsernameKey string `json:"usernameKey,omitempty"`
+}
+
+func (credentials *RootCredentialConfig) validateEitherFromVaultSecretOrFromSecretOrFromRandomSecret() error {
+	count := 0
+	if credentials.RandomSecret != nil {
+		count++
+	}
+	if credentials.Secret != nil {
+		count++
+	}
+	if credentials.VaultSecret != nil {
+		count++
+	}
+	if count != 1 {
+		return errors.New("only one of spec.rootCredentials.vaultSecret or spec.rootCredentials.secret or spec.rootCredentials.randomSecret can be specified")
+	}
+	return nil
+}
+
 func GetFinalizer(instance client.Object) string {
 	return "controller-" + strings.ToLower(instance.GetObjectKind().GroupVersionKind().Kind)
 }
+
+func GetVaultKV2Secret(path string, context context.Context) (*vault.Secret, error) {
```

Comment:
this should be called `GetVaultSecret`. Also I'm pretty sure we already have this function....vaultiutils.readSecret or vaultutils.Read.
While you are at it can you refacrtor read to call readSecret? I should have caught it in the last merge.

Reply:
Made readSecret a public method, replaced GetVaultSecret with it and updated ReadSecret to use new method